File: java/78.subsets.java
/*
 * @lc app=leetcode id=78 lang=java
 *
 * [78] Subsets
 */

// @lc code=start
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> list = new ArrayList<>();
        Arrays.sort(nums); // not necessary
        // result list, templist, targetlist, start
        backtrack(list, new ArrayList<>(), nums, 0);
        return list;
    }

    private void backtrack(List<List<Integer>> list, List<Integer> tempList, int[] nums, int start) {
        // add
        list.add(new ArrayList<>(tempList));
        // for loop start at i = start
        for (int i = start; i < nums.length; i++) {
            // add to the list
            tempList.add(nums[i]);
            backtrack(list, tempList, nums, i + 1);
            // remove last one
            tempList.remove(tempList.size() - 1);
        }
    }
}
// @lc code=end

File: cpp/78.subsets.cpp
/*
 * @lc app=leetcode id=78 lang=cpp
 *
 * [78] Subsets
 */

// @lc code=start

#include <vector>
using namespace std;

class Solution {
   public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> current;

        backtrack(result, current, nums, 0);
        return result;
    }

    void backtrack(vector<vector<int>>& result, vector<int>& current, vector<int>& nums,
                   int start) {
        // base case
        // also add the empty set []
        result.push_back(current);

        // loop
        for (int i = start; i < nums.size(); i++) {
            current.push_back(nums[i]);
            backtrack(result, current, nums, i + 1);
            current.pop_back();
        }
        // example: [1,2,3]
        // [1] -> [1,2] -> [1,2,3] -> [1,3] -> [2] -> [2,3] -> [3]
    }
};
// @lc code=end
